test/socket/test_unix.rb

```diff
@@ -446,6 +442,61 @@ def test_dgram_pair
     s2.close if s2
   end
 
+  def test_stream_pair
+    s1, s2 = UNIXSocket.pair(Socket::SOCK_STREAM)
+    assert_raise(IO::EAGAINWaitReadable) do
+      s1.recv_nonblock(10)
+    end
+
+    s2.send("", 0)
+    s2.send("haha", 0)
+    assert_equal("haha", s1.recv(10))
+    assert_raise(IO::EAGAINWaitReadable) { s1.recv_nonblock(10) }
+
+    buf = "".dup
+    s2.send("BBBBBB", 0)
+    IO.select([s1])
+    rv = s1.recv(100, 0, buf)
+    assert_equal buf.object_id, rv.object_id
+    assert_equal "BBBBBB", rv
+
+    s2.close
+    assert_nil(s1.recv(10))
+  rescue Errno::EPROTOTYPE => error
+    omit error.message
+  ensure
+    s1.close if s1
+    s2.close if s2
+  end
+
+  def test_seqpacket_pair
+    s1, s2 = UNIXSocket.pair(Socket::SOCK_SEQPACKET)
+    assert_raise(IO::EAGAINWaitReadable) do
+      s1.recv_nonblock(10)
+    end
+
+    s2.send("", 0)
+    s2.send("haha", 0)
+    assert_equal(nil, s1.recv(10)) # no way to distinguish empty packet from EOF with SOCK_SEQPACKET
```

Comment:
This I'm not very happy with. The situation is that:

  - DGRAM can send empty packets and has no "connection close". So `recv() == 0` can be assumed to be empty packet.
  - STREAM can't send empty packets and has a "connection close". So `recv() == 0` can be assumed to be EOF.

But  `SEQPACKET` is able to send empty packets AND can send a "connection close". So `recv() == 0` is ambiguous.

I searched for ways to disambiguate but couldn't find any. I think `nil` is better as I really don't see the use case for empty packets with `SEQPACKET`, but I can also keep the old behavior if it's deemed better.

Reply:
The only problem I have with this, is making sure we can disambiguate between an empty packet and a closed connection. Do you feel there is any ambiguity?

Reply:
> Do you think there is a solution?

I don't think there is.

>  is it valid to send an empty packet?

Yes.

Reply:
There is one other option that I'm personally exploring and hope to expose in Ruby 3.3.

In `IO::Buffer`, I'm planning to introduce primitives that more directly follow those of the POSIX model.

Those methods return a length rather than try to interpret the result.

It may be sufficient for your use case.

I think `IO` is like a high level, optionally buffered, stream of data. The fact it captures packet based IO seems more like an afterthought than the main purpose. I personally would like to see us split IO into a "IO::Descriptor" or something similar without any buffering or "stream like interface".

Anyway, that aside, `IO::Buffer` might provide an interface more suitable for your use case.

Reply:
`IO::Buffer` will eventually support `recvmsg` and `sendmsg`.